bin/portal_utils/orbit_interface.py

```diff
@@ -0,0 +1,352 @@
+    ERROR5 = 'ERROR 5: Unknown username:'
+
+    # Handled explicity in deleteUser
+    ERROR6 = 'ERROR 6: Cannot delete user: User is a admin for'
+
+    # Handled explicitly in deleteProject, changeLeader, changeProject
+    ERROR7 = 'ERROR 7: Unknown group name'
+
+    # Handled explicitly in deleteProject
+    ERROR8 = 'ERROR 8: Group/project not deleted because it contains admin(s):'
+
+    # Theoretically could happen from changeProject if I'm trying to move a 
+    # user not created by the portal to a different project. Shouldn't happen.
+    ERROR9 = 'ERROR 9: Cannot move users: different DCs'
+
+    # Malformed LDIF
+    ERROR10 = 'ERROR 10: Missing OU LDIF entry'
+
+    # Malformed LDIF
+    ERROR11 = 'ERROR 11: Missing group name attribute in OU entry'
+
+    # Malformed LDIF
+    ERROR12 = 'ERROR 12: Missing objectClass attribute (organizationalUnit/organizationalRole/organizationalUnit) for'
+
+    # Attempt to create a group without an admin, or with unknown admin 
+    ERROR17 = 'ERROR 17: Missing PI entry'
+
+    # Tried to create a group that already exists.
+    # FIXME: Mostly handled below, but perhaps could be better.
+    ERROR20 = 'ERROR 20: Group exists'
+
+    # Malformed LDIF. Note all users must have an email address.
+    ERROR21 = 'ERROR 21: Missing PI mail:'
+
+    # Malformed LDIF. Note we explicitly require users have an SSH key.
+    ERROR22 = 'ERROR 22: Missing PI ssh public key:'
+
+    # Malformed LDIF.
+    ERROR30 = 'ERROR 30: Missing username (UID)'
+
+    # FIXME: I need to handle this (see comments below). 
+    # This means I tried to add a user to a group that doesn't exist.
+    ERROR31 = 'ERROR 31: Organization does not exist for this user. Missing organization LDIF entry'
+
+    # Malformed LDIF. Not all portal users must have an email address. 
+    ERROR32 = 'ERROR 32: Missing users email address'
+
+    # Malformed LDIF. Note we explicitly require users have an SSH key.
+    ERROR33 = "ERROR 33: Missing user's sshpublickey:"
+
+    ALL = [ERROR1, ERROR2, ERROR3, ERROR4, ERROR5, ERROR5, 
+           ERROR6, ERROR7, ERROR8, ERROR9, ERROR10, ERROR11,
+           ERROR12, ERROR17, ERROR20, ERROR21, ERROR22, 
+           ERROR30, ERROR31, ERROR32, ERROR33]
+
+#  Class to manage interactions with ORBIT through 
+#  Delegated Account Management API
+class ORBIT_Interface:
+
+    def __init__(self, base_orbit_url):
+        self._base_orbit_url = base_orbit_url
+        self._DEVNULL = open(os.devnull, 'w')
+
+
+    # Does given string contain one of the defined errors?
+    # If so, return the error, otherwise, return None
+    def find_error_code(self, str):
+        for cd in CODES.ALL:
+            if str.find(cd) >= 0:
+                return cd;
+        return None
+
+    # Return value from CURL on URL
+    def get_curl_output(self, query_url):
+        raw_curl_output = ""
+        try:
+            raw_curl_output = \
+                subprocess.check_output(["curl", "-k", query_url], 
```

Comment:
Re: timeout, it is a little tricky to do this with subprocess (in python3 they have built in support). Otherwise, you need threads. So I prepended "timeout 20 curl..." to my curl calls and UNIX does it for me. (20 is  variable, of course).